plugins/synaptics-mst/fu-synaptics-mst-firmware.c

```diff
@@ -156,7 +156,10 @@ fu_synaptics_mst_firmware_class_init(FuSynapticsMstFirmwareClass *klass)
 FuFirmware *
 fu_synaptics_mst_firmware_new(void)
 {
-	return FU_FIRMWARE(g_object_new(FU_TYPE_SYNAPTICS_MST_FIRMWARE, NULL));
+	FuSynapticsMstFirmware *self = g_object_new(FU_TYPE_SYNAPTICS_MST_FIRMWARE, NULL);
+	/* default chip family as Tesla */
+	self->family = FU_SYNAPTICS_MST_FAMILY_TESLA;
```

Comment:
I don't think this is a good idea. I think we need to query the firmware file itself to know what generation it is. This seems too easy to flash the "wrong firmware on the wrong hardware" if the value in the other offset is correct.

Reply:
This is just for if caller missed fu_synaptics_mst_firmware_set_family().
How about if I combined fu_synaptics_mst_firmware_set_family() into fu_synaptics_mst_firmware_new()?

Reply:
I think that's working around the problem. Is there really no way to know what generation the firmware is targetting?

Reply:
Two method to determine firmware offset:
1. Check device chip family first, and read corresponding offset(what I implement now).
2. Try to read 3 bytes data of config in firmware file(either start from offset 0x200 or 0x100), it should follow some rule, but we can't 100% sure the content in wrong offset will not fit the rule--that's the risk.

Reply:
Okay, in that case I think we should do *both* -- i.e. explicitly set the family from the device, but try to autodetect the family if unset. Otherwise `fwupdtool firmware-parse` becomes pretty useless.

Reply:
OK, I'll do it later.

Reply:
It should be done in the prepare callback before a write and the write should be rejected without force if they don't match.

Reply:
Added try to auto detect the family if unset in [cdc7651](https://github.com/fwupd/fwupd/pull/6559/commits/cdc7651611c59e6fdc65c974ebb1c11734ce9480)